[Cui Langyuan] iP

clyveycui opened the pull request:
# Duck 🦆 
### What is it?

> "Ducks are cute" - me, while deciding the name of the project, 2021

To put it simply, it is a personal assistant.
Kind of like Jarvis from Ironman, but a lot [_fluffier_](https://hips.hearstapps.com/hmg-prod.s3.amazonaws.com/images/how-to-keep-ducks-call-ducks-1615457181.jpg).
### Features

1. Keep track of your tasks
2. Uhh that's about it

### Todo

- [x] Basic task tracking
- [ ] GUI
- [ ] More features to be decided ~when I stop procrastinating~

### Q & A

- **Why isn't the logo a duck?**

Because ascii is too hard, and I can't find a easy copy and paste solution online. Also 1984

- **But I am more of a chicken person.**

Well, that just sounds like a your problem.

- **Why so many `switch` case statements?**

Let me tell you an old fable, the legend of a certain [yandere dev](https://www.reddit.com/r/ProgrammerHumor/comments/53uhsw/the_code_behind_yandere_simulator/):
```javascript
if (this.StudentID == 6)
{
    if (!false)
    {
        return SubtitleType.TaskGenericLine;
    }
    return SubtitleType.Task6Line;
}
else
{
    if (this.StudentID == 8)
    {
        return SubtitleType.Task8Line;
    }
    if (this.StudentID == 11)
    {
        return SubtitleType.Task11Line;
    }
    if (this.StudentID == 13)
    {
        return SubtitleType.Task13Line;
    }
    if (this.StudentID == 14)
    {
        return SubtitleType.Task14Line;
    }
    if (this.StudentID == 15)
    {
        return SubtitleType.Task15Line;
    }
    if (this.StudentID == 25)
    {
        return SubtitleType.Task25Line;
    }
    if (this.StudentID == 28)
    {
        return SubtitleType.Task28Line;
    }
    if (this.StudentID == 30)
    {
        return SubtitleType.Task30Line;
    }
    if (this.StudentID == 36)
    {
        return SubtitleType.Task36Line;
    }
    if (this.StudentID == 37)
    {
        return SubtitleType.Task37Line;
    }
    if (this.StudentID == 38)
    {
        return SubtitleType.Task38Line;
    }
    if (this.StudentID == 52)
    {
        return SubtitleType.Task52Line;
    }
    if (this.StudentID == 76)
    {
        return SubtitleType.Task76Line;
    }
    if (this.StudentID == 77)
    {
        return SubtitleType.Task77Line;
    }
    if (this.StudentID == 78)
    {
        return SubtitleType.Task78Line;
    }
    if (this.StudentID == 79)
    {
        return SubtitleType.Task79Line;
    }
    if (this.StudentID == 80)
    {
        return SubtitleType.Task80Line;
    }
    if (this.StudentID == 81)
    {
        return SubtitleType.Task81Line;
    }
    return SubtitleType.TaskGenericLine;
}
```
> "Not like switch is any better, but it is what it is." - My friend, passing comment while looking at my code

Commit:
Add Gradle support

clyveycui committed:
Level-1

Week 2 Level-1, greet and echo and exit for Duck

clyveycui committed:
Week 2 Level-2

Added functionality to store user input text in a list.

clyveycui committed:
Week 2 Level-3

added ability to mark task as done

clyveycui committed:
Week 2 Level-4

Added support for 3 types of tasks to replace old task system

clyveycui committed:
Week 2 A-TextUiTesting

Added automatic testing

clyveycui committed:
Week 2 Level-5

Duck now detects basic command line errors and handles it

clyveycui committed:
Week 2 Level-6

Added deletion of tasks from tasklist

clyveycui committed:
Week 2 A-Enums

Enumeration added in previous commits

clyveycui committed:
Reformat switch statements

clyveycui committed:
Fix bugs with event command

clyveycui committed:
Add data folder

clyveycui committed:
Add save task feature to Duke

clyveycui committed:
Change task date from String to LocalDate

clyveycui committed:
Merge branch 'branch-Level-8'

clyveycui committed:
Merge branch level 8 into master

clyveycui committed:
Refactor code to utilize more OOP

clyveycui committed:
Refactor code into packages

clyveycui committed:
Add unit testing for Task package

clyveycui committed:
Fix bug with data folder not being created

clyveycui committed:
Build v0.1-alpha

clyveycui committed:
Add documentation for exception and task

clyveycui committed:
Add documentation for command

clyveycui committed:
Reformat code to adhere to style guide

clyveycui committed:
Merge branch 'branch-A-JavaDoc'

clyveycui committed:
Add find functionality

clyveycui committed:
Merge remote-tracking branch 'origin/add-gradle-support'

clyveycui committed:
Correct mainClassName in build.gradle

clyveycui committed:
Add checkstyle

clyveycui committed:
Format code to follow checkstyle

clyveycui committed:
Add GUI to Duck

clyveycui committed:
Add FXML for GUI

clyveycui committed:
Fix dialog box size to now scale

junwei26 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+     * Creates a new TaskList.
+     */
+    public TaskList() {
+        userList = new ArrayList<Task>(100);
```
Might want to declare 100 as a constant (e.g. `private static final int MAX_ARRAY_SIZE = 100`)

junwei26 commented on src/main/java/duke/DukeStorage.java:
```diff
@@ -0,0 +1,110 @@
+    private static String dataFolder = "data/";
+    private static String partialLoadMessage = "you may have a corrupted/edited save file. Tasks partially loaded";
+    private static String successLoadMessage = "Tasks successfully loaded!";
+
```
Same here, all these constants can be declared as final, you might want to follow the naming convention for constants (e.g. `TASKS_FILE`)

clyveycui replied:
Good points

junwei26 commented on src/main/java/duke/DukeDate.java:
```diff
@@ -0,0 +1,41 @@
+public class DukeDate {
+    private static String dateInputFormat = "d/MM/yyyy";
+    private static String dateOutputFormat = "E, d MMM yyyy";
+
```
These variables can be declared as final since the format isn't gonna change

junwei26 commented on src/main/java/duke/DukeCommandParser.java:
```diff
@@ -0,0 +1,136 @@
+    public static DukeCommand parseCommand(String input) throws DukeCommandException, DukeArgumentException {
+        String[] inputArr = input.split(" ", 2);
+        DukeCommand dc = new CommandGeneric();
+        if (inputArr.length == 1) {
```
I think this is quite subjective but personally I would throw the responsibility of input validation to the `Command `classes as it would simplify the `parseCommand` by alot

clyveycui replied:
That is definitely a good idea, but tbh it will be too much refactoring now... I will think about it.

junwei26 reviewed (commented):
LGTM! No major issues

CrownKira commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,69 @@
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
```
Should this import be removed since it is unused?

clyveycui replied:
Yep, has been removed but not pushed yet

CrownKira commented on src/main/java/duke/DukeCommandParser.java:
```diff
@@ -0,0 +1,136 @@
+import duke.task.Task;
+import duke.task.ToDo;
+
+public class DukeCommandParser {
```
Perhaps it would be nice to include a header comment since it is a public class?

CrownKira replied:
I noticed the same issue in several other places too...

clyveycui replied:
Haven't gotten around to writing docs for every class yet lol

CrownKira commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,24 @@
+
+public class Duke {
+    private static TaskList tList = new TaskList();
+
```
Perhaps you could be more consistent with the line breaks? ie. 1 line separation between field and method. Not a big deal though.

CrownKira replied:
Noticed the same issues in several other places too

clyveycui replied:
Right, will work on that

CrownKira commented on src/main/java/duke/command/CommandGeneric.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    /**
+     * As described in DukeCommand.
```
Perhaps you should start your header comment with a verb? As described in 2103T Java Style Guide 

```
Write the first sentence as a short summary of the method, as Javadoc automatically places it in the method summary table (and index).

    In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)
```

CrownKira replied:
I believe java doc is not needed for overridden methods. So, in this case, you can just remove it. :))

clyveycui replied:
Hmm ok, will remove it

CrownKira reviewed (commented):
LGTM! :) I could only identify some minor issues. One piece of advice I would give is to run `./gradlew checkstyleMain` before pushing. See ya next time! *quack* *quack* *quack*

clyveycui committed:
Change dialogbox color

clyveycui committed:
Adjust code based on peer review

There are several style violations as noted by reviewers.

This causes the code base to not have consistency.

Restyling the code as suggested by peer reviewers:
* Declare final variables where suitable
* Remove extra line breaks
* Delete unnecessary method headers
* Remove magic numbers

This improves the consistency of the code and makes it more readable.

clyveycui committed:
Add assert checks

There are several assumptions in code that will lead to bugs if not met.

Adding assert checks helps to spot these potential bugs.

Let's check the following:
* TaskList
* DukeStorage
* DukeDate

clyveycui committed:
Improve code quality

Code quality was poor in several files, especially in DukeCommandParser.

This made it rather hard to maintain.

Redesigning it such that each command class handles it's own input parsing will make the code more readable.

Let's do just that.

clyveycui committed:
Merge pull request #1 from clyveycui/branch-A-Assertions

Add assert checks

clyveycui committed:
Merge branch 'master' into branch-A-CodeQuality

clyveycui committed:
Update DukeCommandException.java

clyveycui committed:
Merge pull request #2 from clyveycui/branch-A-CodeQuality

Improve code quality

clyveycui committed:
Create github-actions.yml

clyveycui committed:
Merge pull request #3 from clyveycui/branch-A-CI

Create github-actions.yml

clyveycui committed:
Add tagging for tasks

clyveycui committed:
Improve GUI

clyveycui committed:
Add user guide

clyveycui committed:
Add theme to user guide

clyveycui committed:
Set theme jekyll-theme-slate

clyveycui committed:
Add delayed exit for gui
